### PR TITLE
Support returning [versioned] S3Path on `Base.write`

### PR DESCRIPTION
**Pull request by hannahilea**
Addresses https://github.com/JuliaCloud/AWSS3.jl/issues/282 by adding an optional `return_path` flag to `Base.write(fp::S3Path, ...)`. 

Should be unbreaking, as previous return behavior is maintained; I added some tests to show that old return behavior, as well as docstrings for the updated functions.

Does not add support for returning (versioned) path for any of the following: 
- any copy operation (e.g. `s3_copy`)
- any mkdir operation 
- `Base.write(::IO, ...)`

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204817188676381

### CODE REVIEW

**Comment by hannahilea** [comment]
@omus I'm not able to tag you directly, but this is ready for review!

File: README.md [hunk 1]
```diff
@@ -32,5 +32,7 @@ p = S3Path("s3://my.bucket/test1.txt")  # provides an filesystem-like interface
 write(p, "some data")
 
 read(p, byte_range=1:4)  # returns b"some"
+
+p_versioned = write(p, "other data"; return_path=true) # returns path identical to p but with addition of `version`
```
Comment (omus) [inline README.md:36]:
```suggestion
p_versioned = write(p, "other data"; return_path=true) # returns the `S3Path` written to S3 including the new version ID 
```

File: src/AWSS3.jl [hunk 2]
```diff
@@ -919,6 +921,8 @@ s3_purge_versions(a...; b...) = s3_purge_versions(global_aws_config(), a...; b..
 - `metadata::Dict=`; `x-amz-meta-` headers.
 - `tags::Dict=`; `x-amz-tagging-` headers
                  (see also [`s3_put_tags`](@ref) and [`s3_get_tags`](@ref)).
+- `return_raw::Bool=`; when `true`, return un-parsed (raw) `S3.put_object` response
+- `kwargs`; additional kwargs passed through into `S3.put_object`
```
Comment (omus) [inline src/AWSS3.jl:925]: This docstring needs some love. Mentioning this as a reminder to myself

File: src/AWSS3.jl [hunk 3]
```diff
@@ -1033,11 +1040,34 @@ function s3_complete_multipart_upload(
...
+# API Calls
+
+- [`CreateMultipartUpload`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html)
+- [`UploadPart`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html)
+- [`CompleteMultipartUpload`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html)
+
+# Optional Arguments
+- `part_size_mb`: maximum size per uploaded part, in bytes.
+- `return_raw`: when `true`, return un-parsed (raw) `S3.put_object` response
+- `kwargs`: additional kwargs passed through into `s3_upload_part` and `s3_complete_multipart_upload`
```
Comment (omus) [inline src/AWSS3.jl:1061]: I've been trying to put the API calls and permissions at the end of the docstring as most readers don't care.
```suggestion
# Optional Arguments
- `part_size_mb`: maximum size per uploaded part, in bytes.
- `return_raw`: when `true`, return un-parsed (raw) `S3.put_object` response
- `kwargs`: additional kwargs passed through into `s3_upload_part` and `s3_complete_multipart_upload`

# API Calls

- [`CreateMultipartUpload`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html)
- [`UploadPart`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html)
- [`CompleteMultipartUpload`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html)
```

File: src/s3path.jl [hunk 4]
```diff
@@ -651,14 +668,22 @@ function Base.write(
         throw(ArgumentError("Can't write to a specific object version ($(fp.version))"))
     end
 
-    if !multipart || length(content) < MAX_HTTP_BYTES
-        return s3_put(get_config(fp), fp.bucket, fp.key, content)
+    return_raw = return_path
```
Comment (omus) [inline src/s3path.jl:671]:
```suggestion
```

File: src/s3path.jl [hunk 5]
```diff
@@ -651,14 +668,22 @@ function Base.write(
...
+    config = get_config(fp)
+    response = if !multipart || length(content) < MAX_HTTP_BYTES
+        s3_put(config, fp.bucket, fp.key, content; return_raw)
```
Comment (omus) [inline src/s3path.jl:674]:
```suggestion
    response = if !multipart || length(content) < MAX_HTTP_BYTES
        s3_put(config, fp.bucket, fp.key, content; return_raw=true)
```

File: src/s3path.jl [hunk 6]
```diff
@@ -651,14 +668,22 @@ function Base.write(
...
     else
         io = IOBuffer(content)
-        return s3_multipart_upload(
-            get_config(fp), fp.bucket, fp.key, io, part_size_mb; other_kwargs...
+        s3_multipart_upload(
+            config, fp.bucket, fp.key, io, part_size_mb; return_raw, other_kwargs...
```
Comment (omus) [inline src/s3path.jl:678]:
```suggestion
            config, fp.bucket, fp.key, io, part_size_mb; return_raw=true, other_kwargs...
```

File: src/s3path.jl [hunk 7]
```diff
@@ -651,14 +668,22 @@ function Base.write(
...
         )
     end
+
+    if return_path
+        version = get(Dict(response.headers), "x-amz-version-id", nothing)
+        return S3Path(config, fp.bucket, fp.key; version)
+    end
+    return response
```
Comment (omus) [inline src/s3path.jl:686]: I think this keeps things cleaner:
```suggestion
    if return_path
        version = get(Dict(response.headers), "x-amz-version-id", nothing)
        return S3Path(config, fp.bucket, fp.key; version)
    else
        return parse(response)
    end
```

**Review by omus — COMMENTED** [review]
Partial review

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors delegate+

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
bors try

File: src/s3path.jl [hunk 8]
```diff
@@ -651,14 +668,28 @@ function Base.write(
...
+            fp.key,
+            io,
+            part_size_mb;
+            return_raw=true,
+            other_kwargs...,
         )
     end
+
+    if return_path
+        version = get(Dict(response.headers), "x-amz-version-id", nothing)
```
Comment (omus) [inline src/s3path.jl]:
```suggestion
        version = HTTP.header(response.response, "x-amz-version-id", nothing)
```

Comment (hannahilea) [reply src/s3path.jl]: done (needed to add an HTTP dep)

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
test success!

File: test/s3path.jl [hunk 9]
```diff
@@ -689,7 +712,13 @@ function s3path_tests(base_config)
...
 
                 # Create an object which will have versionId set to "null"
-                s3_put(config, b, k, "original")
+                r1 = s3_put(config, b, k, "original"; return_raw)
+                if return_raw
+                    @test isa(r1, AWS.Response)
+                    @test isnothing(HTTP.header(r1.headers, "x-amz-version-id", nothing))
+                else
+                    @test r1 == UInt8[]
+                end
```
Comment (omus) [inline test/s3path.jl:721]:
```suggestion
                r1 = s3_put(config, b, k, "original"; return_raw=true)
                returned_version = HTTP.header(r1.headers, "x-amz-version-id", nothing)
                @test isnothing(returned_version)
```

File: test/s3path.jl [hunk 10]
```diff
@@ -700,12 +729,19 @@ function s3path_tests(base_config)
                 @test versioning_enabled(config, b)
 
                 # Overwrite the original object with a new version
-                s3_put(config, b, k, "new and improved!")
+                r2 = s3_put(config, b, k, "new and improved!"; return_raw)
```
Comment (omus) [inline test/s3path.jl:732]:
```suggestion
                r2 = s3_put(config, b, k, "new and improved!"; return_raw=true)
                returned_version = HTTP.header(r2.headers, "x-amz-version-id", nothing)
                @test !isnothing(returned_version)
```

File: test/s3path.jl [hunk 11]
```diff
@@ -700,12 +729,19 @@ function s3path_tests(base_config)
...
                 @test length(versions) == 2
                 @test versions[1] == "null"
                 @test versions[2] != "null"
+                if return_raw
+                    @test isa(r2, AWS.Response)
+                    @test versions[2] ==
+                        HTTP.header(r2.headers, "x-amz-version-id", nothing)
+                else
+                    @test r2 == UInt8[]
+                end
```
Comment (omus) [inline test/s3path.jl:744]:
```suggestion
                @test versions[2] == returned_version
```

File: test/s3path.jl [hunk 12]
```diff
@@ -736,8 +772,9 @@ function s3path_tests(base_config)
             with_aws_config(config) do
                 # Setup: create a file holding a string `abc`
                 path = S3Path("s3://$(bucket_name)/test_str.txt")
-                write(path, "abc")
+                result = write(path, "abc")
                 @test read(path, String) == "abc"  # Have access to read file
+                @test result == UInt8[]
```
Comment (omus) [inline test/s3path.jl:777]:
```suggestion
                write(path, "abc")
                @test read(path, String) == "abc"  # Have access to read file
```

File: test/awss3.jl [hunk 13]
```diff
@@ -27,31 +27,72 @@ function awss3_tests(base_config)
         @test isempty(s3_get_tags(config, bucket_name))
     end
 
-    @testset "Create Objects" begin
+    @testset "Create Objects" for return_raw in [true, false]
```
[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl:30]: - [x] rolled back

[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl]: - [x] added test

File: test/awss3.jl [hunk 14]
```diff
@@ -77,9 +118,13 @@ function awss3_tests(base_config)
...
         config = assume_testset_role("ReadWriteObject"; base_config)
         xml = "<?xml version='1.0'?><Doc><Text>Hello</Text></Doc>"
-        s3_put(config, bucket_name, "file.xml", xml, "text/xml")
+        result = s3_put(config, bucket_name, "file.xml", xml, "text/xml")
+        @test result == UInt8[]
         @test String(s3_get(config, bucket_name, "file.xml"; raw=true)) == xml
         @test s3_get(config, bucket_name, "file.xml")["Text"] == "Hello"
+
+        result = s3_put(config, bucket_name, "file.xml", xml, "text/xml"; return_raw=true)
+        @test isa(result, AWS.Response)
```
[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl:89]: done

File: test/awss3.jl [hunk 15]
```diff
@@ -230,13 +306,16 @@ function awss3_tests(base_config)
         @test length(collect(s3_list_objects(config, bucket_name))) >= 0
     end
 
-    @testset "Version is empty" begin
+    @testset "Version is empty" for return_raw in [true, false]
```
[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl:272]: done

File: test/awss3.jl [hunk 16]
```diff
@@ -145,8 +192,9 @@ function awss3_tests(base_config)
             ("some.tar.gz", "application/octet-stream"),
             ("data.bz2", "application/octet-stream"),
         ]
-            s3_put(config, bucket_name, k, "x")
+            result = s3_put(config, bucket_name, k, "x"; return_raw)
             @test ctype(k) == t
+            @test isa(result, expected_put_result_type)
         end
     end
```
[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl:162]: done

File: test/awss3.jl [hunk 17]
```diff
@@ -202,15 +274,19 @@ function awss3_tests(base_config)
     #
     # MinIO isn't currently setup with the restrictive prefix required to make the tests
     # fail with "AccessDenied".
-    is_aws(base_config) && @testset "Restricted Prefix" begin
+    is_aws(base_config) && @testset "Restricted Prefix" for return_raw in [true, false]
```
[the remark this answers is not in the document]
Comment (hannahilea) [reply test/awss3.jl:233]: done

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
bors try

File: src/s3path.jl [hunk 18]
```diff
@@ -651,14 +670,41 @@ function Base.write(
         throw(ArgumentError("Can't write to a specific object version ($(fp.version))"))
     end
 
-    if !multipart || length(content) < MAX_HTTP_BYTES
-        return s3_put(get_config(fp), fp.bucket, fp.key, content)
+    supported_return_values = [:parsed, :response, :path]
```
Comment (omus) [inline src/s3path.jl:673]:
```suggestion
    supported_return_values = (:parsed, :response, :path)
```

File: test/s3path.jl [hunk 19]
```diff
@@ -225,9 +234,31 @@ end
...
+        @test isa(result, AWS.Response)
+    end
+end
+
+function test_write_returns(ps::PathSet)
+    teststr = "Test string"
+    @test write(ps.quux, teststr) == UInt8[]
+    @test write(ps.quux, teststr; returns=:parsed) == UInt8[]
+    @test isa(write(ps.quux, teststr; returns=:response), AWS.Response)
+    @test isa(write(ps.quux, teststr; returns=:path), S3Path)
```
Comment (omus) [inline test/s3path.jl:260]:
```suggestion
    @test write(ps.quux, teststr; returns=:response) isa AWS.Response
    @test write(ps.quux, teststr; returns=:path) isa S3Path
```

File: test/s3path.jl [hunk 20]
```diff
@@ -610,19 +645,25 @@ function s3path_tests(base_config)
 
     # `s3_list_versions` gives `SignatureDoesNotMatch` exceptions on Minio
     if is_aws(base_config)
-        @testset "S3Path versioning" begin
+        @testset "S3Path versioning" for parse_response in [true, false]
```
Comment (omus) [inline test/s3path.jl:648]:
```suggestion
        @testset "S3Path versioning" begin
```

File: test/s3path.jl [hunk 21]
```diff
@@ -610,19 +645,25 @@ function s3path_tests(base_config)
...
             config = assume_testset_role("S3PathVersioningTestset"; base_config)
 
             s3_enable_versioning(config, bucket_name)
-            key = "test_versions"
-            s3_put(config, bucket_name, key, "data.v1")
-            s3_put(config, bucket_name, key, "data.v2")
+            key = "test_versions_$(parse_response)"
+            r1 = s3_put(config, bucket_name, key, "data.v1"; parse_response)
+            r2 = s3_put(config, bucket_name, key, "data.v2"; parse_response)
```
Comment (omus) [inline test/s3path.jl:654]:
```suggestion
            key = "test_versions"
            r1 = s3_put(config, bucket_name, key, "data.v1"; parse_response=false)
            r2 = s3_put(config, bucket_name, key, "data.v2"; parse_response=false)
            rv1 = HTTP.header(r1.headers, "x-amz-version-id", nothing)
            rv2 = HTTP.header(r2.headers, "x-amz-version-id", nothing)
```

File: test/s3path.jl [hunk 22]
```diff
@@ -610,19 +645,25 @@ function s3path_tests(base_config)
...
             listed_versions = s3_list_versions(config, bucket_name, key)
             versions = [d["VersionId"] for d in reverse!(listed_versions)]
 
             v1, v2 = first(versions), last(versions)
+            if parse_response
+                @test r1 == r2 == UInt8[]
+            else
+                @test v1 == HTTP.header(r1.headers, "x-amz-version-id", nothing)
+                @test v2 == HTTP.header(r2.headers, "x-amz-version-id", nothing)
+            end
```
Comment (omus) [inline test/s3path.jl:666]:
```suggestion
            @test v1 == rv1
            @test v2 == rv2
```

File: test/s3path.jl [hunk 23]
```diff
@@ -689,7 +730,9 @@ function s3path_tests(base_config)
                 @test !versioning_enabled(config, b)
 
                 # Create an object which will have versionId set to "null"
-                s3_put(config, b, k, "original")
+                r1 = s3_put(config, b, k, "original"; parse_response=false)
+                returned_version = HTTP.header(r1.headers, "x-amz-version-id", nothing)
+                @test isnothing(returned_version)
```
Comment (omus) [inline test/s3path.jl:735]:
```suggestion
                rv1 = HTTP.header(r1.headers, "x-amz-version-id", nothing)
                @test isnothing(rv1)
```

File: test/s3path.jl [hunk 24]
```diff
@@ -700,12 +743,15 @@ function s3path_tests(base_config)
                 @test versioning_enabled(config, b)
 
                 # Overwrite the original object with a new version
-                s3_put(config, b, k, "new and improved!")
+                r2 = s3_put(config, b, k, "new and improved!"; parse_response=false)
+                returned_version = HTTP.header(r2.headers, "x-amz-version-id", nothing)
+                @test !isnothing(returned_version)
```
Comment (omus) [inline test/s3path.jl:748]:
```suggestion
                r2 = s3_put(config, b, k, "new and improved!"; parse_response=false)
                rv2 = HTTP.header(r2.headers, "x-amz-version-id", nothing)
                @test !isnothing(rv2)
```

File: test/s3path.jl [hunk 25]
```diff
@@ -700,12 +743,15 @@ function s3path_tests(base_config)
...
 
                 versions = list_version_ids(config, b, k)
                 @test length(versions) == 2
                 @test versions[1] == "null"
                 @test versions[2] != "null"
+                @test versions[2] == returned_version
```
Comment (omus) [inline test/s3path.jl:754]:
```suggestion
                @test versions[2] == rv2
```

File: test/s3path.jl [hunk 26]
```diff
@@ -225,9 +234,31 @@ end
 function test_large_write(ps::PathSet)
     teststr = repeat("This is a test string!", round(Int, 2e5))
     @testset "large write/read" begin
```
Comment (omus) [inline test/s3path.jl:236]:
```suggestion
function test_multipart_write(ps::PathSet)
    teststr = repeat("This is a test string!", round(Int, 2e5))
    @testset "multipart write" begin
```

File: test/s3path.jl [hunk 27]
```diff
@@ -225,9 +234,31 @@ end
...
+    @testset "large write/read, return response" begin
+        result = write(ps.quux, teststr; part_size_mb=1, multipart=true, returns=:response)
+        @test read(ps.quux, String) == teststr
+        @test isa(result, AWS.Response)
+    end
+end
+
+function test_write_returns(ps::PathSet)
+    teststr = "Test string"
+    @test write(ps.quux, teststr) == UInt8[]
```
Comment (omus) [inline test/s3path.jl:257]: Needs an testset: `@testset "Write returns" begin`

File: src/s3path.jl [hunk 28]
```diff
@@ -634,6 +634,24 @@ function Base.read(fp::S3Path; byte_range=nothing)
...
+
+# Optional Arguments
+- `multipart`: when `true`, uploads data via [`s3_multipart_upload`](@ref) for `content`
+  greater than `part_size_mb` bytes; when false, or when `content` is shorter
+  than `part_size_mb`, uploads data via [`s3_put`](@ref).
+- `part_size_mb`: when `multipart=true`, sets maximum length of partitioned data (in bytes).
+- `returns`: type of returned object, one of `:response` (the `AWS.Response`
+  of the internal [`s3_put`](@ref)); `:parsed` (default; the parsed `AWS.Response`); or
+  `path` (the [`S3Path`](@ref) written to, including a `version` if versioning is
+  enabled for `fp.bucket`).
```
Comment (omus) [inline src/s3path.jl:652]:
```suggestion
- `returns`: determines the result returned by the function: `:response` (the AWS API
  response), :parsed` (default; the parsed AWS API response), or `:path` (the newly created
  [`S3Path`](@ref), including its `version` when versioning is enabled for the bucket).
```

**Review by omus — COMMENTED** [review]
Should be good after this review

File: src/AWSS3.jl [hunk 29]
```diff
@@ -45,6 +45,7 @@ using Dates
 using EzXML
 using FilePathsBase
 using FilePathsBase: /, join
+using HTTP
```
Comment (omus) [inline src/AWSS3.jl:48]:
```suggestion
using HTTP: HTTP
```

**Comment by hannahilea** [comment]
bors try

**Comment by hannahilea** [comment]
hrm. when i do a `bors try` i get that "already running a review" note, but i don't see where the review is running.... at any rate, @omus, ready to go (i think).

**Comment by hannahilea** [comment]
bors try

**Review by omus — APPROVED** [review]
Looks great!

**Pull request merged**